#162162982 Comment on highlighted text

aayebare opened the pull request:
#### What does this PR do?
- Enable users to highlight any given text in the article's body.
- Enable users to comment on the highlighted text.

#### Description of Task to be completed?
- Implement highlighting a particular text in an article
- Implement commenting on the highlighted text

#### How should this be manually tested?
- On the local machine, pull this branch using  `git pull origin ft-comment-on-highlighted-text-162162982` command
- Then use the command ```git checkout ft-comment-on-highlighted-text-162162982``` to switch to this branch
- Open your editor and start a new project workspace.
- Create, activate and install the dependencies on a virtual environment
- Create a Postgres database called `ahdb` and make migrations by running:
        * `python manage.py make migrations `
        * `python manage.py migrate` 
- Run the server using python manage.py runserver
- Using swagger docs on (`http://127.0.0.1:8000/docs/`), Comment on created articles using the `POST api/{slug}/comments/` endpoint
- In the structure of the comment, you can highlight a text from an article by declaring a start index position and an end index position. **Note.**  This is not a requirement to make a comment to an article
- The format of the JSON object is as;
```
{
  "comment": {
        "body": "This is a parent comment",
        "start_index_position": 13,
        "end_index_position": 22
                     }
}
```
**Note.** Ensure to use integer values
- You can use the`GET api/{slug}/comments/` endpoint to view the comments on a particular article

#### What are the relevant pivotal tracker stories?
#162162982

malep2007 commented on authors/apps/comments/utils.py:
```diff
@@ -0,0 +1,36 @@
+            start_index_position = comment.get('start_index_position', 0)
+            end_index_position = comment.get('end_index_position', 0)
+            try:
+                isinstance(start_index_position, int)
```
This section of code is redundant in this part. It will either return a true/ False and not catch any exceptions. This is not how exceptions are handled. The only section which could possibly raise an exception is where the `selected_text` gets a `None` object.

malep2007 replied:
it could be simplified even further like this
```python
if 'start_index_position' and 'end_index_position' in comment:
        try:
            start_index_position = int(comment.get('start_index_position', 0))
            end_index_position = int(comment.get('end_index_position', 0))
            selected_text = str(article_slug.body[start_index_position:end_index_position])
            comment['selected_text'] = selected_text
        except TypeError:
            raise ValidationError({
                "message": "Please use integer values to select a start and end position and ensure you highlight the text 
                you want"
            })
    return comment
```
This is by far not the best implementation but you can surely figure out something that works.

malep2007 commented on authors/apps/tests/test_comment_on_article.py:
```diff
@@ -122,3 +120,19 @@ def test_delete_comment(self):
+        slug = self.slugger()
+        url = f'/api/{slug}/comments/'
+        response = self.client.post(url, data=highlight_comment_2, format="json")
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)   
```
**NEW LINES PLEASE**

malep2007 reviewed (changes requested):
Please work on the feedback

katunold reviewed (changes requested):
Just work on the feedback from Ephraim and you are good to go. otherwise, well done

billkabanga commented:
The first 3 lines in the tests with helper methods could be Refactored into one helper method that you can call in the test methods. 
Good work done though. Implement all the feedback and all will be okay.

Commit:
feat(highlight): users highlight and comment on an article

- users can highlight sections of a text in an article
- users can comment on the highlighted text

[Finishes #162162982]
